### PR TITLE
ReturnTypeChangedProblem

### LINKED ISSUE #1628: MethodWithReturnTypeNotFoundProblem
Followup of https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/1541#issuecomment-681879475

`MethodWithReturnTypeNotFoundProblem` is a special case of `SymbolNotFoundProblem`. MethodWithReturnTypeNotFoundProblem occurs when Linkage Checker finds a method with name and argument types but the return type is different. This specific class enables us to print detailed message such as

> com.Foo's method R method(A, B, C) is not found. Return type is different.

### PR DESCRIPTION
**Pull request by suztomo**
Fixes #1628

### CODE REVIEW

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java [hunk 1]
```diff
@@ -274,20 +274,40 @@ private boolean problemFilter(LinkageProblem linkageProblem) {
...
+      boolean methodWithDifferentReturnTypeFound = false;
       for (JavaClass javaClass : typesToCheck) {
         for (Method method : javaClass.getMethods()) {
-          if (method.getName().equals(methodName)
-              && method.getSignature().equals(symbol.getDescriptor())) {
-            if (!isMemberAccessibleFrom(javaClass, method, sourceClassName)) {
-              return Optional.of(
-                  new InaccessibleMemberProblem(sourceClassFile, targetClassFile, symbol));
+          if (method.getName().equals(methodName)) {
+            String signatureInTargetClass = method.getSignature();
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java:282]: actualSignature

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java [hunk 2]
```diff
@@ -274,20 +274,40 @@ private boolean problemFilter(LinkageProblem linkageProblem) {
...
+            String signatureLookingFor = symbol.getDescriptor();
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java:283]: expectedSignature

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java [hunk 3]
```diff
@@ -274,20 +274,40 @@ private boolean problemFilter(LinkageProblem linkageProblem) {
...
+                return Optional.of(
+                    new InaccessibleMemberProblem(sourceClassFile, targetClassFile, symbol));
+              }
+              // The method is found and accessible. Returning no error.
+              return Optional.empty();
+            } else {
+              String argumentTypeInTarget = parseArgumentTypeParts(signatureInTargetClass);
+              String argumentTypeLookingFor = parseArgumentTypeParts(signatureLookingFor);
+              if (argumentTypeInTarget.equals(argumentTypeLookingFor)) {
+                // Not returning result yet, because there can be other supertype that has the exact
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java:295]: "another supertype" or "other supertypes"

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java [hunk 4]
```diff
@@ -274,20 +274,40 @@ private boolean problemFilter(LinkageProblem linkageProblem) {
           Iterables.concat(
               getClassHierarchy(targetJavaClass),
               Arrays.asList(targetJavaClass.getAllInterfaces()));
+
+      boolean methodWithDifferentReturnTypeFound = false;
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java:278]: returnTypeChanged

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/MethodWithReturnTypeNotFoundProblem.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+package com.google.cloud.tools.opensource.classpath;
+
+import javax.annotation.Nullable;
+
+/**
+ * The {@code sourceClass} references the {@code methodSymbol}, but the {@code
+ * methodSymbol.getClassBinaryName} has the method with different return type.
+ */
+class MethodWithReturnTypeNotFoundProblem extends LinkageProblem {
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/MethodWithReturnTypeNotFoundProblem.java:25]: ReturnTypeChangedProblem

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java (hunk 4 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java]: Updated.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java (hunk 1 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java]: Updated.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java (hunk 2 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java]: Updated.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java (hunk 3 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java]: Updated to another supertype.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/MethodWithReturnTypeNotFoundProblem.java (hunk 5 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/MethodWithReturnTypeNotFoundProblem.java]: Updated.

**Comment by suztomo** [comment]
@elharo PTAL

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java [hunk 6]
```diff
@@ -274,20 +274,39 @@ private boolean problemFilter(LinkageProblem linkageProblem) {
...
+            String expectedSignature = symbol.getDescriptor();
+            if (actualSignature.equals(expectedSignature)) {
+              if (!isMemberAccessibleFrom(javaClass, method, sourceClassName)) {
+                return Optional.of(
+                    new InaccessibleMemberProblem(sourceClassFile, targetClassFile, symbol));
+              }
+              // The method is found and accessible. Returning no error.
+              return Optional.empty();
+            } else {
+              String argumentTypeInTarget = parseArgumentTypeParts(actualSignature);
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java:292]: actualArgumentType

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java [hunk 7]
```diff
@@ -274,20 +274,39 @@ private boolean problemFilter(LinkageProblem linkageProblem) {
...
+              String argumentTypeLookingFor = parseArgumentTypeParts(expectedSignature);
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java:293]: expectedArgumentType

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java [hunk 8]
```diff
@@ -274,20 +274,39 @@ private boolean problemFilter(LinkageProblem linkageProblem) {
...
+              if (argumentTypeInTarget.equals(argumentTypeLookingFor)) {
+                // Not returning result yet, because there can be another supertype that has the
+                // exact method that match the name, argument types, and return type.
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java:296]: matches

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java (hunk 6 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java]: Updated

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java (hunk 7 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java]: Updated

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java (hunk 8 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java]: Fixed.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ReturnTypeChangedProblem.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+ * actualTypeName}.
+ */
+class ReturnTypeChangedProblem extends LinkageProblem {
+  public ReturnTypeChangedProblem(
+      ClassFile sourceClass,
+      @Nullable ClassFile targetClass,
+      MethodSymbol expectedMethodSymbol,
+      String actualTypeName) {
+    super(
+        "is not found. The expected return type does not match actual type " + actualTypeName,
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ReturnTypeChangedProblem.java:33]: Now it tells the actual type.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ReturnTypeChangedProblem.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+import javax.annotation.Nullable;
+
+/**
+ * The {@code sourceClass} references the {@code methodSymbol}, but the {@code
+ * methodSymbol.getClassBinaryName} has the method with a different return type {@code
+ * actualTypeName}.
+ */
+class ReturnTypeChangedProblem extends LinkageProblem {
+  public ReturnTypeChangedProblem(
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ReturnTypeChangedProblem.java:27]: non-public

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ReturnTypeChangedProblem.java (hunk 9 above)
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ReturnTypeChangedProblem.java:33]: match the

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ReturnTypeChangedProblem.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.opensource.classpath;
+
+import javax.annotation.Nullable;
+
+/**
+ * The {@code sourceClass} references the {@code methodSymbol}, but the {@code
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ReturnTypeChangedProblem.java:22]: {@code methodSymbol} does not exist below. {@code expectedMethodSymbol} perhaps?

Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ReturnTypeChangedProblem.java]: Fixed.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ReturnTypeChangedProblem.java (hunk 10 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ReturnTypeChangedProblem.java]: Fixed.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ReturnTypeChangedProblem.java (hunk 9 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ReturnTypeChangedProblem.java]: Updated.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java [hunk 12]
```diff
@@ -274,20 +275,41 @@ private boolean problemFilter(LinkageProblem linkageProblem) {
...
+              String actualArgumentType = parseArgumentTypeParts(actualSignature);
+              String expectedArgumentType = parseArgumentTypeParts(expectedSignature);
+              if (actualArgumentType.equals(expectedArgumentType)) {
+                // Not returning result yet, because there can be another supertype that has the
+                // exact method that matches the name, argument types, and return type.
+                changedReturnTypeName = Utility.methodSignatureReturnType(actualSignature);
+              }
             }
-            // The method is found and accessible. Returning no error.
-            return Optional.empty();
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java:286]: This has been moved to line 290.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ReturnTypeChangedProblem.java [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+ * The {@code sourceClass} references the {@code expectedMethodSymbol}, but the {@code
+ * expectedMethodSymbol.getClassBinaryName} has the method with a different return type ({@code
+ * actualTypeName}).
+ */
+class ReturnTypeChangedProblem extends LinkageProblem {
+  ReturnTypeChangedProblem(
+      ClassFile sourceClass,
+      @Nullable ClassFile targetClass,
+      MethodSymbol expectedMethodSymbol,
+      String actualTypeName) {
```
[the remark this answers is not in the document]
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ReturnTypeChangedProblem.java]: Updated.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java [hunk 14]
```diff
@@ -274,20 +275,41 @@ private boolean problemFilter(LinkageProblem linkageProblem) {
           Iterables.concat(
               getClassHierarchy(targetJavaClass),
               Arrays.asList(targetJavaClass.getAllInterfaces()));
+
+      String changedReturnTypeName = null;
```
[the remark this answers is not in the document]
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java]: Updated.

**Review by suztomo — COMMENTED** [review]
@elharo PTAL.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java [hunk 15]
```diff
@@ -304,6 +326,13 @@ private boolean problemFilter(LinkageProblem linkageProblem) {
     }
   }
 
+  /** Returns the argument type parts from {@code methodDescriptor}. */
+  private static String parseArgumentTypeParts(String methodDescriptor) {
+    // E.g., '(Ljava/lang/String;)V' => '(Ljava/lang/String;)'
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java:331]: I don't follow this comment. This is an identity transformation?

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java [hunk 16]
```diff
@@ -304,6 +326,13 @@ private boolean problemFilter(LinkageProblem linkageProblem) {
     }
   }
 
+  /** Returns the argument type parts from {@code methodDescriptor}. */
+  private static String parseArgumentTypeParts(String methodDescriptor) {
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java:330]: would this somehow be different from parseArgumentTypes?

**Review by elharo — APPROVED** [review]

**Pull request merged**